planner/core/rule_join_reorder.go

```diff
@@ -27,37 +27,63 @@ import (
 	"golang.org/x/exp/slices"
 )
 
+type joinMethodHint struct {
+	preferredJoinMethod uint
+	joinMethodHintInfo  *tableHintInfo
+}
+
 // extractJoinGroup extracts all the join nodes connected with continuous
 // Joins to construct a join group. This join group is further used to
 // construct a new join order based on a reorder algorithm.
 //
 // For example: "InnerJoin(InnerJoin(a, b), LeftJoin(c, d))"
 // results in a join group {a, b, c, d}.
 func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction,
-	otherConds []expression.Expression, joinTypes []*joinTypeWithExtMsg, hintInfo []*tableHintInfo, hasOuterJoin bool) {
+	otherConds []expression.Expression, joinTypes []*joinTypeWithExtMsg, joinOrderHintInfo []*tableHintInfo, joinMethodHintInfo map[int]*joinMethodHint, hasOuterJoin bool) {
```

Comment:
Too many return values, can we merge them into a structure like `joinGroupResult{}`?